### PR TITLE
Add user roles and secure admin level endpoints

### PR DESCRIPTION
**Pull request by miikahyttinen**
- add user and admin roles
- secure admin level endpoints
- save info about role to firebase in custom claims object

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -12,6 +12,7 @@ val kotlinVersion = "1.7.10"
 val jacksonVersion = "2.13.3"
 val springBootVersion = "2.7.2"
 val junitVersion = "5.9.0"
+val springSecurityVersion = "5.7.3"
```
Comment (LeadingMoominExpert) [inline build.gradle.kts:15]: Where is this used? Or just leftover from something?

File: src/main/kotlin/three/consulting/epoc/config/SecurityConfig.kt [hunk 2]
```diff
@@ -3,28 +3,48 @@ package three.consulting.epoc.config
...
 
 @Profile("default")
-@EnableWebSecurity
+@Component
+@EnableGlobalMethodSecurity(
+    prePostEnabled = true
+)
 class SecurityConfig() {
 
+    val authoritiesClaimName = "role"
```
Comment (LeadingMoominExpert) [inline src/main/kotlin/three/consulting/epoc/config/SecurityConfig.kt:22]: In kotlin it's very common to wrap constant values inside a companion object for clarity and define them as private, and write in caps snake case constants if needed
```kotlin
private companion object {
    const val AUTHORITIES_CLAIM_name = "role"
}
```

File: src/main/kotlin/three/consulting/epoc/dto/FirebaseUserDTO.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package three.consulting.epoc.dto
+
+import javax.validation.constraints.NotBlank
+
+data class FirebaseUserDTO(
+    @field:NotBlank val role: String,
+    @field:NotBlank val firebaseUid: String
+)
```
Comment (LeadingMoominExpert) [inline src/main/kotlin/three/consulting/epoc/dto/FirebaseUserDTO.kt:8]: This feels like a redundant class, since the fields already exist in `EmployeeDTO` can't you just use that? This is already used in `EmployeeController` so feels natural it would just be the employee given as the request

File: src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                .build()
+            val firebaseApp = FirebaseApp.initializeApp(options)
+            return FirebaseAuth.getInstance(firebaseApp)
+        } catch (e: java.lang.Exception) {
+            logger.error(e) { "Could not set up Firebase" }
+        }
+        return null
+    }
+
+    private val firebaseAuth = firebaseAuth()
```
Comment (LeadingMoominExpert) [inline src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt:34]: Instead of defining the auth method and calling it here, why not just define it as a bean inside FirebaseConfig?
```kotlin
@Profile("default")
@Configuration
class FirebaseConfig(
    private val firebaseService: FirebaseService
) : InitializingBean {
    override fun afterPropertiesSet() {
    ...
    }

    @Bean
    fun firebaseAuth() =
        try {
            val options = FirebaseOptions.builder()
                .setCredentials(GoogleCredentials.getApplicationDefault())
                .build()
            val firebaseApp = FirebaseApp.initializeApp(options)
            return FirebaseAuth.getInstance(firebaseApp)
        } catch (e: Exception) {
            logger.error(e) { "Could not set up Firebase" }
            throw e
        }
```
This way firebaseauth is initialized in the configuration when everything is injected to spring context and it's grouped inside a configuration and not a service.

Also there's no need to use `java.lang.Exception`. Kotlin has it's own shorthands so you can just use `Exception` without importing anything

File: src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import three.consulting.epoc.dto.EmployeeDTO
+import three.consulting.epoc.dto.FirebaseUserDTO
+import three.consulting.epoc.entity.Employee
+import three.consulting.epoc.repository.EmployeeRepository
+
+private val logger = KotlinLogging.logger {}
+
+@Service
+class FirebaseService(
+    private val employeeRepository: EmployeeRepository
```
Comment (LeadingMoominExpert) [inline src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt:19]: If you configure the firebaseAuth() method as a bean, you can just add in the constructor here after the repository
```kotlin
private val firebaseAuth: FirebaseAuth
```

File: src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        } else {
+            employee.firebaseUid = firebaseUid
+            firebaseAuth!!.setCustomUserClaims(firebaseUid, customClaims)
+            EmployeeDTO(employeeRepository.save(employee))
+        }
+    }
+
+    fun updateFirebaseUserRole(firebaseUserDTO: FirebaseUserDTO) {
+        val customClaims = mapOf("role" to firebaseUserDTO.role)
+        firebaseAuth!!.setCustomUserClaims(firebaseUserDTO.firebaseUid, customClaims)
```
Comment (LeadingMoominExpert) [inline src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt:56]: Also if you define firebaseAuth as a bean which either returns a valid FirebaseAuth or throws an exception, you can get rid of this dangerous double bang operator

**Review by LeadingMoominExpert — CHANGES_REQUESTED** [review]
Some kotlin and spring specific stuff that need to be worked on to make everything cleaner and more coherent, but looks pretty good.

Biggest thing imo is the `FirebaseUserDTO` class which I don't think is needed at all. All the data is present in an employee.

File: build.gradle.kts (hunk 1 above)
Comment (miikahyttinen) [reply build.gradle.kts:15]: Yep, at some point I thought I need it.

File: src/main/kotlin/three/consulting/epoc/dto/FirebaseUserDTO.kt (hunk 3 above)
Comment (miikahyttinen) [reply src/main/kotlin/three/consulting/epoc/dto/FirebaseUserDTO.kt:8]: I would like to require _role_ and _firebaseUid_ values with the endpoint - these are optional in _EmployeeDTO_ class. Also there are some field that are unnecesary in this particular operation. However, we could create a new FirebaseController. On the other hand, if it has only one endpoint it might be unnecessary.

Comment (LeadingMoominExpert) [reply src/main/kotlin/three/consulting/epoc/dto/FirebaseUserDTO.kt:8]: They can stay as optional for now, but they should be mandatory for employee as well. The problem is we need to run the current migrations to production first. We can make new migrations making the uid and role mandatory in a future pr

**Review by LeadingMoominExpert — CHANGES_REQUESTED** [review]

**Review by LeadingMoominExpert — APPROVED** [review]
Good job 👏🏻  Just rename the pr to not have an issue number and maybe reword the title a bit and then merge
